<a href="https://colab.research.google.com/github/hpdnnr7/CSc448/blob/main/Hope_Dunner_Assignment_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd 
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import datasets
from sklearn.preprocessing import PolynomialFeatures
from sklearn.model_selection import cross_val_score

## Part 1. Loading the dataset

In [ ]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv("https://raw.githubusercontent.com/profmcnich/example_notebook/main/science_data_large.csv")
# Output the first 15 rows of the data
dataset.head(15)

Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04

In [ ]:
# Display a summary of the table information (number of datapoints, etc.)
df = pd.DataFrame(dataset)
print("Summary of the basic information about this DataFrame and its data:")
print(df.info())
display(df.describe())

Summary of the basic information about this DataFrame and its data:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB
None


Temperature °C     Mols KCL     Size nm^3
count     1000.000000  1000.000000  1.000000e+03
mean       500.500000   471.530000  5.086111e+05
std        288.819436   288.482872  4.474838e+05
min          1.000000     1.000000  1.611429e+01
25%        250.750000   226.750000  1.298267e+05
50%        500.500000   459.500000  3.827182e+05
75%        750.250000   710.250000  7.603211e+05
max       1000.000000  1000.000000  1.972127e+06

## Part 2. Splitting the dataset

In [ ]:
# Take the pandas dataset and split it into our features (X) and label (y)
features = df[["Temperature °C", "Mols KCL"]]
label = df["Size nm^3"]
X, y = features, label
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=.10, random_state=42)
X_train, X_test, y_train, y_test

(     Temperature °C  Mols KCL
 716              95       499
 351             243       384
 936              47       211
 256             495       497
 635              82       421
 ..              ...       ...
 106             439        40
 270             665       245
 860             471       620
 435             829       259
 102             636       388
 
 [900 rows x 2 columns],
      Temperature °C  Mols KCL
 521             100       541
 737             635       668
 740             799       665
 660             966       871
 411             785       595
 ..              ...       ...
 436             421       725
 764             431       860
 88              983       433
 63              580       817
 826              24        89
 
 [100 rows x 2 columns],
 716    103064.31430
 351    193753.02860
 936     21670.02857
 256    505027.40000
 635     75092.02857
            ...     
 106     40433.71429
 270    335545.00000
 860    600674.85710
 435    44128

## Part 3. Perform a Linear Regression

In [ ]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model
model = LinearRegression().fit(X_train, y_train)
model.score(X_train, y_train)
# Report on the score for that model, in your own words (markdown, not code) explain what the score means
# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX

0.8610082946223787

In [ ]:
model.predict(np.array([[3,5]]))

/usr/local/lib/python3.9/site-packages/sklearn/base.py:441: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([-401629.56501086])

In [ ]:
model.coef_

array([ 866.14641337, 1032.69506649])

In [ ]:
model.intercept_

-409391.47958340764

The score 0.86, represents an 86% accuracy rate on the equation of a slime. 

The function below represents the h(x) including the coefficents and intercept from the model

$$ h(x) = 866.14641337 * x_0 + 1032.69506649 * x_1 -409391.47958340764$$

Sample equation: $E = mc^2$

## Part 4. Use Cross Validation

In [ ]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data

# Report on their finding and their significance

In [ ]:
X_train.shape, y_train.shape

((900, 2), (900,))

In [ ]:
X_test.shape, y_test.shape

((100, 2), (100,))

In [ ]:
clf = cross_val_score(model,X_train, y_train)
print(clf)

[0.86226163 0.81982226 0.88938198 0.86663176 0.85729958]


In [ ]:
clf.mean(), clf.std()

(0.8590794402339516, 0.022493119105236794)

In [ ]:
print("0.86 accuracy with a standard deviation of 0.02")

0.86 accuracy with a standard deviation of 0.02


## Part 5. Using Polynomial Regression

In [ ]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(2)
X_train = poly.fit_transform(X_train)
X_test = poly.fit_transform(X_test)

model = model.LinearRegression()

#polynomial regression to linear regression

model.fit(X_train, y_train)
# Report on the metrics and output the resultant equation as you did in Part 3.


AttributeError: 'LinearRegression' object has no attribute 'LinearRegression'

In [ ]:
poly_model.score(X_train, y_train)

In [ ]:
poly_model.coef_

In [ ]:
poly_model.intercept_

In [ ]:
poly_model.predict(X_test)

$$ h(x) = (0.00 * 1 +  1.20 * a -1.27 * b + 1.26 * a^2 +
        2.00 * ab + 2.85 * b^2)  $$